src/core/Akka.Persistence/Fsm/PersistentFSM.cs

```diff
@@ -5,96 +5,405 @@
+                    else if (@event is StateChangeEvent)
                     {
-                        base.ApplyState(upcomingState.Using(nextData));
-
-                        if (upcomingState.AfterTransitionHandler != null)
-                        {
-                            upcomingState.AfterTransitionHandler(upcomingState.StateData);
-                        }
+                        applyStateOnLastHandler();
                     }
                 });
             }
         }
     }
+
+    public static class PersistentFSM
+    {
+        /// <summary>
+        /// IFsmState interface, makes possible for simple default serialization by conversion to String
+        /// </summary>
+        public interface IFsmState
+        {
+            string Identifier { get; }
+        }
+
+        /// <summary>
+        /// Persisted on state change
+        /// </summary>
+        internal class StateChangeEvent : IMessage
+        {
+            /// <summary>
+            /// Initializes a new instance of the <see cref="StateChangeEvent"/> class.
+            /// </summary>
+            /// <param name="stateIdentifier">FSM state identifier.</param>
+            /// <param name="timeout">FSM state timeout.</param>
+            public StateChangeEvent(string stateIdentifier, TimeSpan? timeout)
+            {
+                StateIdentifier = stateIdentifier;
+                Timeout = timeout;
+            }
+
+            /// <summary>
+            /// FSM state identifier.
+            /// </summary>
+            public string StateIdentifier { get; }
+
+            /// <summary>
+            /// FSM state timeout.
+            /// </summary>
+            public TimeSpan? Timeout { get; }
+        }
+
+        /// <summary>
+        /// FSM state and data snapshot
+        /// </summary>
+        internal class PersistentFSMSnapshot<TD>
+        {
+            /// <summary>
+            /// Initializes a new instance of the <see cref="PersistentFSMSnapshot{TD}"/> class.
+            /// </summary>
+            /// <param name="stateIdentifier">FSM state identifier.</param>
+            /// <param name="data">FSM state data</param>
+            /// <param name="timeout">FSM state timeout.</param>
+            public PersistentFSMSnapshot(string stateIdentifier, TD data, TimeSpan? timeout)
+            {
+                StateIdentifier = stateIdentifier;
+                Data = data;
+                Timeout = timeout;
+            }
+
+            /// <summary>
+            /// FSM state identifier.
+            /// </summary>
+            public string StateIdentifier { get; }
+
+            /// <summary>
+            /// FSM state data.
+            /// </summary>
+            public TD Data { get; }
+
+            /// <summary>
+            /// FSM state timeout.
+            /// </summary>
+            public TimeSpan? Timeout { get; }
+        }
+
+        public class State<TS, TD, TE>
```

Comment:
Probably not, this class has a mix of custom generic types, `IReadOnlyList<object>` and Action delegates, none of which is guaranteed to provide structural equality.